src/Interpreters/ColumnAliasesVisitor.cpp

```diff
@@ -0,0 +1,98 @@
+#include <Interpreters/ColumnAliasesVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Interpreters/RequiredSourceColumnsVisitor.h>
+#include <Interpreters/addTypeConversionToAST.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/ASTSelectWithUnionQuery.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTSubquery.h>
+#include <Parsers/ASTAlterQuery.h>
+#include <Parsers/ASTInsertQuery.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/queryToString.h>
+
+namespace DB
+{
+
+bool ColumnAliasesMatcher::needChildVisit(const ASTPtr & node, const ASTPtr &)
+{
+    if (const auto * f = node->as<ASTFunction>())
+    {
+        /// "lambda" visit children itself.
+        if (f->name == "lambda")
+            return false;
+    }
+
+    return !(node->as<ASTTableExpression>()
+            || node->as<ASTSubquery>()
+            || node->as<ASTArrayJoin>()
+            || node->as<ASTSelectQuery>()
+            || node->as<ASTSelectWithUnionQuery>());
+}
+
+void ColumnAliasesMatcher::visit(ASTPtr & ast, Data & data)
+{
+    auto aa = queryToString(ast);
+    // If it's select query, only replace filters.
+    if (auto * query = ast->as<ASTSelectQuery>())
+    {
+        if (query->where())
+             Visitor(data).visit(query->refWhere());
+        if (query->prewhere())
+             Visitor(data).visit(query->refPrewhere());
+
+        return;
+    }
+
+    if (auto * node = ast->as<ASTFunction>())
+    {
+        visit(*node, ast, data);
+        return;
+    }
+
+    if (auto * node = ast->as<ASTIdentifier>())
+    {
+        visit(*node, ast, data);
+        return;
+    }
+}
+
+void ColumnAliasesMatcher::visit(ASTFunction & node, ASTPtr & /*ast*/, Data & data)
+{
+    /// Do not add formal parameters of the lambda expression
+    if (node.name == "lambda")
+    {
+        Names local_aliases;
+        for (const auto & name : RequiredSourceColumnsMatcher::extractNamesFromLambda(node))
```

Comment:
Syntax looks too tricky, surrounding `for` with brackes looks better (and less  error prone)

Reply:
fixed